banner/models.py

```diff
@@ -1,189 +1,40 @@
-    paths = models.TextField(
-        null=True,
-        blank=True,
-        help_text="""A list of regular expressions that match the path. This \
-is used to find a banner to render when using a banner proxy. One entry per \
-line. If you are unfamiliar with regular expressions then enter the relative \
-URL to an item, eg. /post/my-post."""
-    )
-
-
-class CodeBanner(Banner):
-    code = models.TextField(
-        help_text='The full HTML/Javascript code snippet to be \
-embedded for this banner.'
-    )
-
-
-class BaseImageBanner(Banner):
-    """For every different aspect ratio of image banner a subclass is required.
-    Use this class as base."""
-    url = models.CharField(
-        max_length='256',
-        verbose_name='URL',
-        help_text='URL (internal or external) to which this banner will link.'
-    )
-    dfp_slot_name = models.CharField(
-        max_length=256,
-        null=True,
-        blank=True,
-        help_text="""A combination of network code and ad unit as provided \
-by Google, eg. /1234/travel. Used to track clicks."""
-    )
-    dfp_ad_id = models.CharField(
-        max_length=64,
-        null=True,
-        blank=True,
-        help_text="The advertisement id. Used to track clicks."
-    )
-
-    class Meta:
-        abstract = True
-
-    def get_target_url(self):
-        if self.dfp_slot_name and self.dfp_ad_id:
-            url = "%s?slot_name=%s&ad_id=%s&url=%s" % \
-            (
-                reverse('banner-dfp-click-proxy'), self.dfp_slot_name,
-                self.dfp_ad_id, self.url
-            )
-            return url
-        return self.url
-
-    @property
-    def dfp_tracking_url(self):
-        # http://support.google.com/dfp_sb/bin/answer.py?hl=en&answer=1651549
-        # tracking pixel section.
-        # xxx: I was expecting an ad id but there is none in the example.
-        if self.dfp_slot_name:
-            url = "http://pubads.g.doubleclick.net/gampad/ad?iu=%s&sz=1x1&t=&c=%s" % \
-                (self.dfp_slot_name, randint(0, 2000000000))
-        else:
-            url = ''
-        return url
+from django.utils.translation import ugettext_lazy as _
 
+from jmbo.models import Image, ModelBase
+from link.models import Link
 
-class ImageBanner(BaseImageBanner):
-    pass
+from banner.styles import BANNER_STYLE_CLASSES
 
 
-class BaseDFPBanner(Banner):
-    slot_name = models.CharField(
-        max_length=256,
-        help_text="""A combination of network code and ad unit as provided \
-by Google, eg. /1234/travel."""
+class Banner(ModelBase):
+    """Base class for all banners"""
+    link = models.ForeignKey(
+        Link, help_text=_("Link to which this banner should redirect."),
+        blank=True, null=True
     )
-    width = models.PositiveIntegerField()
-    height = models.PositiveIntegerField()
-    targeting_key = models.CharField(
-        max_length=128,
-        help_text="Eg. 'interests'."
+    background_image = models.OneToOneField(
```

Comment:
Rather just make use of the image inline. The style/template combo should dictate what it tries to do with images. eg: It could be a banner with background images that need to cycle too.

Reply:
Or use the modelbase base image as the background.

Reply:
I like the idea that the style/template should dictate how to handle images. 
The reason I didn't go with that was because I didn't want to make too many assumptions.
i.e., avoiding a situation where image uses is inferred from the position of the image in the return queryset